File: java/83.remove-duplicates-from-sorted-list.java
/*
 * @lc app=leetcode id=83 lang=java
 *
 * [83] Remove Duplicates from Sorted List
 */
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode deleteDuplicates(ListNode head) {
        if (head == null || head.next == null) return head;
        ListNode prev = head, ptr = head.next;
        while (ptr != null) {
            if (prev.val != ptr.val) prev = ptr;
            else prev.next = ptr.next;
            ptr = ptr.next;
        }
        return head;
    }
}

File: go/83.remove-duplicates-from-sorted-list.go
/*
 * @lc app=leetcode id=83 lang=golang
 *
 * [83] Remove Duplicates from Sorted List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func deleteDuplicates(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}
	ahead := &ListNode{0, head}
	prev, cur, next := ahead, head, head.Next
	lastSeen := cur.Val - 1
	for {
		for lastSeen == cur.Val {
			prev.Next = next
			if next == nil {
				break
			}
			cur, next = next, next.Next
		}
		lastSeen = cur.Val
		if next == nil {
			break
		}
		prev, cur, next = cur, next, next.Next
	}
	return ahead.Next
}
// @lc code=end

